### PR TITLE
Implement RFC 1461

### PR DESCRIPTION
**Pull request by sfackler**
I have these tagged as stable in 1.9, so this shouldn't merge until the 1.8 beta's cut.

### CODE REVIEW

**Comment by sfackler** [comment]
Works on Windows now.

I ended up backing out the addition of `TcpStream::{set_,}keepalive`. There's actually two operations going on here - enabling keepalive and changing the interval from the system default. In addition, you can't load the interval on Windows, so `TcpStream::keepalive` doesn't even work there (it'll always return `Ok(None)`).

We should probably split the keepalive functionality into two methods to mirror how things work, and possibly move the interval accessor to a unix-only extension trait, but that should be worked out in the net2 crate.

File: src/libstd/net/tcp.rs [hunk 1]
```diff
@@ -180,6 +180,89 @@ impl TcpStream {
...
+    /// [link]: #tymethod.set_ttl
+    #[stable(feature = "net2_mutators", since = "1.9.0")]
+    pub fn ttl(&self) -> io::Result<u32> {
+        self.0.ttl()
+    }
+
+    /// Sets the value for the `IPV6_V6ONLY` option on this socket.
+    ///
+    /// If this is set to `true` then the socket is restricted to sending and
+    /// receiving IPv6 packets only. In this case two IPv4 and IPv6 applications
```
Comment (cybergeek94) [inline src/libstd/net/tcp.rs]: > In this case two IPv4 and IPv6 applications can bind the same port at the same time

This sounds somewhat misleading. Wouldn't that be one of each?

Comment (sfackler) [inline src/libstd/net/tcp.rs]: Fixed

File: src/libstd/net/tcp.rs [hunk 2]
```diff
@@ -969,4 +1113,48 @@ mod tests {
...
+        let listener = t!(TcpListener::bind(&addr));
+
+        t!(listener.set_nonblocking(true));
+        t!(listener.set_nonblocking(false));
+
+        let stream = t!(TcpStream::connect(&("localhost", addr.port())));
+
+        t!(stream.set_nonblocking(true));
+        t!(stream.set_nonblocking(false));
+    }
```
Comment (alexcrichton) [inline src/libstd/net/tcp.rs]: Could you add a read here to make sure it doesn't block?

File: src/libstd/sys/common/net.rs [hunk 3]
```diff
@@ -40,8 +40,7 @@ pub fn setsockopt<T>(sock: &Socket, opt: c_int, val: c_int,
 }
 
 pub fn getsockopt<T: Copy>(sock: &Socket, opt: c_int,
-                       val: c_int) -> io::Result<T> {
-    unsafe {
+                       val: c_int) -> io::Result<T> { unsafe {
```
Comment (alexcrichton) [inline src/libstd/sys/common/net.rs]: maybe a mixup here?

**Comment by alexcrichton** [comment]
Thanks @sfackler! 1.8 has been branched, so feel free to r=me with a few nits addressed

**Pull request merged**